File: problems/30-floyd-warshall/002-path-reconstruction-negative-cycles/java/Solution.java
import java.util.*;

public class Solution {
    static final long INF = Long.MAX_VALUE / 4;
    private long[][] dist;
    private int[][] next;

    public void build(int n, int[][] edges) {
        dist = new long[n][n]; next = new int[n][n];
        for (int i=0;i<n;i++) Arrays.fill(dist[i], INF);
        for (int i=0;i<n;i++) { dist[i][i]=0; next[i][i]=i; }
        for (int[] e: edges) if (e[2] < dist[e[0]][e[1]]) { dist[e[0]][e[1]]=e[2]; next[e[0]][e[1]]=e[1]; }
        for (int k=0;k<n;k++) for (int i=0;i<n;i++) for (int j=0;j<n;j++)
            if (dist[i][k] < INF && dist[k][j] < INF && dist[i][j] > dist[i][k]+dist[k][j]) {
                dist[i][j]=dist[i][k]+dist[k][j]; next[i][j]=next[i][k];
            }
    }

    public boolean hasNegativeCycle() {
        for (int i=0;i<dist.length;i++) if (dist[i][i] < 0) return true;
        return false;
    }

    public List<Integer> path(int from, int to) {
        if (next[from][to] == 0 && from != to) return Collections.emptyList();
        List<Integer> result = new ArrayList<>();
        result.add(from);
        while (from != to) { from = next[from][to]; result.add(from); if (result.size() > dist.length + 1) return Collections.emptyList(); }
        return result;
    }
}

File: problems/30-floyd-warshall/002-path-reconstruction-negative-cycles/python/solution.py
class FloydWarshall:
    INF = 10**30

    def build(self, n, edges):
        self.dist = [[self.INF] * n for _ in range(n)]
        self.next = [[None] * n for _ in range(n)]
        for i in range(n):
            self.dist[i][i] = 0
            self.next[i][i] = i
        for u, v, w in edges:
            if w < self.dist[u][v]:
                self.dist[u][v] = w
                self.next[u][v] = v
        for k in range(n):
            for i in range(n):
                for j in range(n):
                    if self.dist[i][k] < self.INF and self.dist[k][j] < self.INF:
                        cand = self.dist[i][k] + self.dist[k][j]
                        if cand < self.dist[i][j]:
                            self.dist[i][j] = cand
                            self.next[i][j] = self.next[i][k]

    def has_negative_cycle(self):
        return any(self.dist[i][i] < 0 for i in range(len(self.dist)))

    def path(self, u, v):
        if self.next[u][v] is None:
            return []
        result = [u]
        while u != v:
            u = self.next[u][v]
            result.append(u)
            if len(result) > len(self.dist) + 1:
                return []
        return result
